### ISSUE TITLE
Can not resovle optional requirements for other platform.

### ISSUE THREAD

**Issue by songww** [sdispater/poetry#164]
The package [docker](https://github.com/docker/docker-py) has required `pypiwin32` for windows platform extras deps. But I got an `SolverProblemError`, when I add docker at linux.

```
⇒  poetry add docker --extras tls            
Using version ^3.3 for docker

Updating dependencies
Resolving dependencies... (0.1s)
                                                                                       
[SolverProblemError]                                                    
Because docker (3.3.0) depends on both pypiwin32 (220) and pypiwin32 (219), docker   
is forbidden.                                                                        
So, because no versions of docker match >3.3,<4.0                                    
 and optools depends on docker (^3.3), version solving failed.
```

https://github.com/docker/docker-py/blob/master/setup.py#L29

**Comment by songww** [comment]
⇒  poetry --version                               
Poetry 0.10.1

**Comment by songww** [comment]
platform is linux. but, the selected version is win32 platform.

**Comment by bertjwregeer** [comment]
```
            "idna (>=2.0.0); extra == 'tls'",
            "cryptography (>=1.3.4); extra == 'tls'",
            "pyOpenSSL (>=0.14); extra == 'tls'",
            "pypiwin32 (==220); sys_platform == \"win32\" and python_version >= \"3.6\"",
            "pypiwin32 (==219); sys_platform == \"win32\" and python_version < \"3.6\"",
            "backports.ssl-match-hostname (>=3.5); python_version < \"3.5\"",
            "ipaddress (>=1.0.16); python_version < \"3.3\"",
            "docker-pycreds (>=0.2.3)",
            "websocket-client (>=0.32.0)",
            "six (>=1.4.0)",
            "requests (!=2.18.0,>=2.14.2)"
```

It looks like the resolver is unable to figure out the `python_version` on the `requires_dist`.

While your platform may be Linux, `poetry` tries to resolve all dependencies for all platforms and locks them, so that is why it is complaining about windows only dependencies.

**Comment by sdispater** [comment]
The dependency resolver will pick up dependencies as long as they are compatible with what is defined in the `pyproject.toml`.

In this case, if you have a `python` requirement that supports Python < 3.6 and >=3.6 then the resolver will pick up both `pypiwin32` versions which leads to the conflict you are seeing.

I will try to find a way to make it accept different versions as long as they have different, non-conflicting, requirements.

**Comment by songww** [comment]
some version in lockfile is wrong too. like `pymongo==3.6.1-py3.6-win32`.

**Comment by sdispater** [comment]
@songww This is not the same problem as this one. Can you open a separate issue with steps to reproduce?

**Comment by songww** [comment]
And there has another problem.
I have specified python version gte 3.6, like `python = ">=3.6"`.
```
⇒  poetry add oss2                        
Creating virtualenv poetry-demo-py3.6 in /home/songww/.cache/pypoetry/virtualenvs
Using version ^2.5 for oss2

Updating dependencies
Resolving dependencies... (17.7s)
                                                                                       
[SolverProblemError]                                                    
Because no versions of aliyun-python-sdk-core match >2.6.2,<2.7.0 || >2.7.0,<2.7.2   
|| >2.7.2,<2.8.0 || >2.8.0,<2.8.1 || >2.8.1,<2.8.2 || >2.8.2                         
 and aliyun-python-sdk-core (2.6.2) requires Python <3, aliyun-python-sdk-core is f  
orbidden.                                                                            
And because aliyun-python-sdk-core (2.7.0) requires Python <3, aliyun-python-sdk-co  
re is forbidden.                                                                     
And because aliyun-python-sdk-core (2.7.2) requires Python <3                        
 and aliyun-python-sdk-core (2.8.0) requires Python <3, aliyun-python-sdk-core is f  
orbidden.                                                                            
And because aliyun-python-sdk-core (2.8.1) requires Python <3                        
 and aliyun-python-sdk-core (2.8.2) requires Python <3, aliyun-python-sdk-core is f  
orbidden.                                                                            
Because no versions of oss2 match >2.5,<3.0                                          
 and oss2 (2.5.0) depends on aliyun-python-sdk-core (>=2.6.2), oss2 (>=2.5,<3.0) re  
quires aliyun-python-sdk-core (>=2.6.2).                                             
Thus, oss2 is forbidden.                                                             
So, because poetry-demo depends on oss2 (^2.5), version solving failed.              
                                                                                       
add [-D|--dev] [--git GIT] [--path PATH] [-E|--extras EXTRAS] [--optional] [--python PYTHON] [--platform PLATFORM] [--allow-prereleases] [--dry-run] [--] <name> (<name>)...
```
If I keep python version as *: `python = "*"`, `oss2` can be installed correctly.

**Comment by sdispater** [comment]
The original issue should be fixed in the latest `0.11.0a3` prerelease.

@songww Your other issue should also be resolved.
